### ISSUE TITLE
Highly correlated data within each chunk

### ISSUE THREAD

**Issue by EmmaRocheteau** [YerevaNN/mimic3-benchmarks#78]
I was digging into the BatchGen class in mimic3models.length_of_stay.utils and I noticed that the _generator() method seems to only pick out data from a few patients at a time, rather a set of shuffled examples taken from the whole training set. For example, if I have a batch_size of 8, then the chunk will either be 8192 (batch_size*1024) in size (or less if we are reaching the end of the full epoch). To take the particular chunk that I'm looking at as an example, it only contains 91 patients, because they fill the 8192 sample quota for the chunk. Doesn't this cause a big problem for the network updates? You're going to have highly correlated data (repeated data of varying lengths) within the chunks, especially if there are unusually long stays within that chunk. Some patients have length of stays that are thousands of hours.

**Comment by niuguy** [comment]
Are you saying the train_listfile is not shuffled?  I checked create_length_of_stay.py which generate the train files for LOS. They do have `    if partition == "train":
        random.shuffle(xty_triples)` at line 73. According to what Ian Goodfellow said in this post [https://www.quora.com/Does-the-order-of-training-data-matter-when-training-neural-networks#
](https://www.quora.com/Does-the-order-of-training-data-matter-when-training-neural-networks#)

> As long as the data has been shuffled, everything should work OK. Different random orderings will perform slightly differently from each other but this will be a small factor that does not matter much.

So I guess it should be all right? Although you can still try and compare different random orderings.

**Comment by Harhro94** [comment]
Here are two important notes:
- All training data generators are created with `shuffle=True`.
- Looking at [this line](https://github.com/YerevaNN/mimic3-benchmarks/blob/master/mimic3models/length_of_stay/utils.py#L45) we see that every time a generator makes a complete pass over the samples, it will shuffle the samples before starting the next pass.

**Comment by EmmaRocheteau** [comment]
I don't think I was clear enough when I asked the original question sorry! The issue is not whether or not the _patients_ are presented in a random order (it is of course a good thing that they are shuffled differently for each pass). It is about the examples being _shuffled across patients_ rather than clustered by patient. Your code picks out random patients without replacement and then puts all of those examples (from the same patient) in the same chunk. The samples from the same patient will be very highly correlated (in fact when you compare any two samples from the same patient, the longer one will contain a complete copy of the shorter one).
